book/src/template_expansion.md

```diff
@@ -0,0 +1,387 @@
+    fn render_into(
+        &self,
+        writer: &mut (impl ::std::fmt::Write + ?Sized),
+    ) -> ::askama::Result<()> {
+        let x = 12;
+        ::askama::Result::Ok(())
+    }
+    // I skipped some internal `askama` impl items to keep this as simple
+    // as possible.
+}
+
+impl ::std::fmt::Display for YourType {
+    #[inline]
+    fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+        ::askama::Template::render_into(self, f).map_err(|_| ::std::fmt::Error {})
+    }
+}
+```
+
+For simplicity, we will now only keep the content of the `askama::Template::render_into`
+function from now on.
+
+## Text content
+
+If you have "text content" (for example HTML) in your template:
+
+```html
+<h1>{{ title }}</h1>
+```
+
+It will generate it like this:
+
+```rust
+writer
+    .write_fmt(
+        format_args!(
+            "<h1>{0}</h1>",
+            &::askama::MarkupDisplay::new_unsafe(&(self.title), ::askama::Html),
+        ),
+    )?;
+::askama::Result::Ok(())
+```
+
+## Variables
+
+### Variables creation
+
+If you create a variable in your template, it will be created in the generated
+Rust code as well. For example:
+
+```jinja
+{% set x = 12 %}
+{% let y = x + 1 %}
+```
+
+will generate:
+
+```rust
+let x = 12;
+let y = x + 1;
+::askama::Result::Ok(())
+```
+
+### Variables usage
+
+By default, variables will be fields from the type on which the `askama::Template`
+trait is implemented:
+
+```jinja
+{{ y }}
+```
+
+will generate:
+
+```rust
+writer
+    .write_fmt(
+        format_args!(
+            "{0}",
+            &::askama::MarkupDisplay::new_unsafe(&(self.y), ::askama::Html),
+        ),
+    )?;
+::askama::Result::Ok(())
+```
+
+This is why if the variable is undefined, it won't work with askama and why
+we can't check if a variable is defined or not.
+
+However, you can still access constants and statics by using paths. Let's say
+you have in your Rust code:
+
+```rust
+const FOO: u32 = 0;
+```
+
+Then you can use them in your template by referring to them with a path:
+
+```jinja
+{{ crate::FOO }}{{ super::FOO }}
+```
+
+It will generate:
+
+```rust
+writer
+    .write_fmt(
+        format_args!(
+            "{0}{1}",
+            &::askama::MarkupDisplay::new_unsafe(&(crate::FOO), ::askama::Html),
+            &::askama::MarkupDisplay::new_unsafe(&(super::FOO), ::askama::Html),
+        ),
+    )?;
+::askama::Result::Ok(())
+```
+
+You can also access items from your current type as well using `Self::`, it'll
+use the same logic.
+
+## Control blocks
+
+### if/else
+
+There is an interesting twist for `if/else` conditions. Let's check this code:
+
+```jinja
+{% if x == "a" %}
+gateau
+{% else %}
+tarte
+{% endif %}
+```
+
+It will generate:
+
+```rust
+if *(&(self.x == "a") as &bool) {
+    writer.write_str("gateau")?;
+} else {
+    writer.write_str("tarte")?;
+}
+::askama::Result::Ok(())
+```
+
+Very much as expected except for the `&(self.x == "a") as &bool`. It allows to
```

Comment:
It would be nice to provide a little explanation of why this is generated like this... I don't remember but `git blame` could probably tell us.

Reply:
Gonna look around.